### PR TITLE
Enhance auto-tune module

### PR DESCRIPTION
**Pull request by yiliu30**
## Type of Change

Enhancement
API changed or not: None

## Description

Enhance auto-tune module's internal implementation; no functionality changes.
- [x] Introduce `TuningParam` to handle the complex tunable type, such as a list of lists.
- [x] Separate the `Sampler` and `ConfigSet` from `ConfigLoader`
- [x] Enhance the `BaseConfig`'s expand 
- [x] discard `timeout` from `TuningConfig` 

## How has this PR been tested?

Pre-CI

## Dependency Change?
- [pydantic](https://github.com/pydantic/pydantic) [MIT](https://github.com/pydantic/pydantic?tab=MIT-1-ov-file#readme)(Need add it to BoM!!)

### CODE REVIEW

File: neural_compressor/common/tuning_param.py [hunk 1]
```diff
@@ -0,0 +1,91 @@
...
+        Class FakeAlgoConfig(BaseConfig):
+            '''Fake algo config.'''.
+
+            params_list = [
+                ...
+                # * For complex tunable types, like a list of lists,
+                # * developers need to create the `TuningParam` explicitly.
+                TuningParam("complex_attr", tunable_type=List[List])
+                # * For simple tunable types, like a list of int, giving the param name is enough.
+                "simple_attr"
```
Comment (yiliu30) [inline neural_compressor/common/tuning_param.py:43]: Hi @mengniwang95, to handle complicated tunable types, such as a list of lists, I would like to introduce this mechanism. Please have a look :).

File: neural_compressor/common/tuning_param.py [hunk 2]
```diff
@@ -0,0 +1,100 @@
...
+            params_list = [
+                ...
+                # For simple tunable types, like a list of int, giving
+                # the param name is enough. `BaseConfig` class will
+                # create the `TuningParam` implicitly.
+                "simple_attr"
+
+                # For complex tunable types, like a list of lists,
+                # developers need to create the `TuningParam` explicitly.
+                TuningParam("complex_attr", tunable_type=List[List])
```
Comment (yiliu30) [inline neural_compressor/common/tuning_param.py:46]: Hi @mengniwang95, to handle complicated tunable types, such as a list of lists, I would like to introduce this mechanism. Please have a look :).

File: neural_compressor/common/tuning_param.py [hunk 3]
```diff
@@ -0,0 +1,100 @@
...
+
+                # The default parameter level is `ParamLevel.OP_LEVEL`.
+                # If the parameter is at a different level, developers need
+                # to specify it explicitly.
+                TuningParam("model_attr", level=ParamLevel.MODEL_LEVEL)
```
Comment (yiliu30) [inline neural_compressor/common/tuning_param.py:51]: Hi @yuwenzho, to handle parameters with specific levels, such as model level, I would like to introduce this mechanism. Please have a look :).

Comment (yiliu30) [reply neural_compressor/common/tuning_param.py:51]: The non-tuning parameter with specific level is still not being handled. Fix it in the subsequent PRs.

**Review by mengniwang95 — APPROVED** [review]

**Review by chensuyue — APPROVED** [review]

**Pull request merged**